### PR TITLE
Doing the loss reduction in foundry instead of in the loss functions.

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
This PR gives us more flexibility to reduce losses in foundry in custom ways.

We see that the changes do not affect the MFU or the convergence for 125M and 7B models.

Screenshot 2024-04-01 at 4 58 49 PM

Screenshot 2024-04-02 at 9 17 43 AM

Memory consumption is also similar:

Screenshot 2024-04-02 at 9 21 09 AM
Screenshot 2024-04-02 at 9 21 38 AM

### CODE REVIEW

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -1015,9 +1017,18 @@ def forward(self, batch: MutableMapping) -> CausalLMOutputWithPast:
...
-        return self.loss_fn(outputs.logits.view(-1, outputs.logits.size(-1)),
-                            targets.view(-1))
+        targets = self.get_targets(batch).view(-1)
+        losses = self.loss_fn(outputs.logits.view(-1, outputs.logits.size(-1)),
+                              targets)
+
+        if torch.all(targets == -100):
+            raise ValueError(
+                'All targets are -100. This might be due to all tokens in a sequence being padding tokens.'
+            )
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:1027]:
```suggestion
            loss = losses.sum()
```

Should we return 0 in this case?
maybe log a warning, but Error doesn't make sense.

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -1015,9 +1017,18 @@ def forward(self, batch: MutableMapping) -> CausalLMOutputWithPast:
 
     def loss(self, outputs: CausalLMOutputWithPast,
              batch: Mapping) -> torch.Tensor:
-        targets = self.get_targets(batch)
-        return self.loss_fn(outputs.logits.view(-1, outputs.logits.size(-1)),
-                            targets.view(-1))
+        targets = self.get_targets(batch).view(-1)
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:1020]: The only place targets is used is in `(targets != self.loss_fn.ignore_index).sum()`.

Do we need to view it as -1 for that?

(I'm a bit scared of destroying batch dim unless its NEEDED)

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 1 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/modeling_mpt.py:1027]: This was done to preserve functionality. The code before this change would result in a NaN being returned as loss, due to dividing by 0. A better way to capture that would be raise an error. That is why I raise this error.

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 2 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/modeling_mpt.py:1020]: There is only one other place the targets is being used in this function, and that also uses `.view(1)`

Screenshot 2024-04-01 at 5 12 21 PM

Comment (dakinggg) [reply llmfoundry/models/mpt/modeling_mpt.py:1020]: @vchiley CE loss expects the flattened targets, right?

Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:1020]: Just trying to future proof the code.
I'd argue that we should pass it into the loss fn as `targets.view(1)` but don't destroy the batch info unless needed.
(what if in the future someone wants the batch info for something else)

Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:1020]: i guess ppl can undo this change in the future, I'm just not sure what the point of including it now is 🤷‍♂️

Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/modeling_mpt.py:1020]: Makes sense, let me update the PR.

**Review by vchiley — APPROVED** [review]

**Review by vchiley — APPROVED** [review]

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -984,7 +984,8 @@ def __init__(
                 from flash_attn.losses.cross_entropy import \
                     CrossEntropyLoss as FusedCrossEntropyLoss
 
-                self.loss_fn = FusedCrossEntropyLoss(ignore_index=-100)
+                self.loss_fn = FusedCrossEntropyLoss(ignore_index=-100,
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:987]: Outside the scope of this pr but should we make expose `ignore_index` in the config (default `-100`)?
cc @dakinggg

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 4]
```diff
@@ -1016,8 +1018,15 @@ def forward(self, batch: MutableMapping) -> CausalLMOutputWithPast:
...
         targets = self.get_targets(batch)
-        return self.loss_fn(outputs.logits.view(-1, outputs.logits.size(-1)),
-                            targets.view(-1))
+        losses = self.loss_fn(outputs.logits.view(-1, outputs.logits.size(-1)),
+                              targets.view(-1))
+
+        if torch.all(targets == self.loss_fn.ignore_index):
+            loss = losses.sum()
+        else:
+            loss = losses.sum() / (targets != self.loss_fn.ignore_index).sum()
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:1027]: Outside the scope of this PR, but should we enable expose loss fn `reduction` in the config (default `'mean'`)?

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 3 above)
Comment (dakinggg) [reply llmfoundry/models/mpt/modeling_mpt.py:987]: I don't think so. Everyone uses -100, not sure why anyone would want to change this

Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:987]: resolving cuz outside the scope of this PR

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 4 above)
Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:1027]: resolving cuz outside the scope of this PR
